File: app/src/main/java/com/kaciula/archiman/presentation/screen/main/UserAdapter.java
package com.kaciula.archiman.presentation.screen.main;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.kaciula.archiman.R;
import com.kaciula.archiman.domain.model.User;
import com.kaciula.archiman.presentation.util.GenericRecyclerAdapter;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class UserAdapter extends GenericRecyclerAdapter<User, UserAdapter.ViewHolder> {

    private final MainContract.Presenter presenter;

    public UserAdapter(Context ctx, List<User> items, MainContract.Presenter presenter) {
        super(ctx, items);
        this.presenter = presenter;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View convertView = inflater.inflate(R.layout.item_user, parent, false);
        return new ViewHolder(convertView);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        User user = getItem(position);
        holder.username.setText(user.name());
        holder.view.setOnClickListener(holder);
    }

    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        @BindView(R.id.username)
        TextView username;

        View view;

        public ViewHolder(View view) {
            super(view);
            this.view = view;
            ButterKnife.bind(this, view);
        }

        @Override
        public void onClick(View view) {
            int adapterPosition = getAdapterPosition();
            if (adapterPosition != RecyclerView.NO_POSITION) {
                User user = getItem(adapterPosition);
                presenter.onClickUser(user);
            }
        }
    }
}

File: app/src/main/java/com/kaciula/archiman/presentation/screens/home/UserAdapter.kt
package com.kaciula.archiman.presentation.screens.home

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.kaciula.archiman.R
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.item_user.*

class UserAdapter(
    var items: List<UserViewModel>,
    private val presenter: HomeContract.Presenter
) : RecyclerView.Adapter<UserAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val convertView =
            LayoutInflater.from(parent.context).inflate(R.layout.item_user, parent, false)
        return ViewHolder(convertView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val user = items[position]
        holder.bind(user)
    }

    override fun getItemCount(): Int = items.size

    inner class ViewHolder(
        override val containerView: View
    ) : RecyclerView.ViewHolder(containerView), LayoutContainer {

        fun bind(user: UserViewModel) {
            tvUsername.text = user.name
            containerView.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    presenter.onClickUser(items[adapterPosition])
                }
            }
            btnDetails.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    presenter.onClickUserDetails(items[adapterPosition])
                }
            }
        }
    }
}
